File: backend/src/Services/BuildingService.cs
using API.Entities;
using API.Interfaces.Repositories;
using API.Types;

namespace API.Services;

public class BuildingService(IBuildingRepository buildingRepository) {

    public async Task<Page<Building>> GetAll(string filter, int page, int limit) {

        Page<Building> buildings = await buildingRepository.GetAll(filter, page, limit);

        return buildings;

    }

    public async Task<Building?> GetBuildingById(int id) {

        Building? building = await buildingRepository.GetById(id);

        return building;

    }

    public async Task<Page<Building>> GetBuildingsByManager(int managerId, string filter, int page, int limit) {

        Page<Building> buildings = await buildingRepository.GetByManagerId(managerId, filter, page, limit);

        return buildings;

    }

    public async Task<Building?> CreateBuilding(int managerId, string address) {

        Building building = new() {
            ManagerId = managerId,
            Address = address
        };

        try {
            return await buildingRepository.Create(building);
        } catch {
            return null;
        }

    }

    public async Task<Building?> UpdateBuilding(int id, int? managerId, string? address) {

        Building? building = await buildingRepository.GetById(id);

        if(building == null) {
            return null;
        }

        building.ManagerId = managerId ?? building.ManagerId;
        building.Address = address ?? building.Address;
        building.Manager = null;

        try {
            return await buildingRepository.Update(building);
        } catch {
            return null;
        }

    }

    public async Task<Building?> DeleteBuilding(int id) {

        Building? building = await buildingRepository.GetById(id);

        if(building == null) {
            return null;
        }

        try {
            return await buildingRepository.Delete(building);
        } catch {
            return null;
        }

    }

}

File: frontend/src/services/building.service.ts
import { APIError } from "@/models/api-error.models";
import { Building } from "@/models/building.model";
import { Page } from "@/models/page";
import { Service } from "@/services/service";
import { optional } from "@/types/optional";

export class BuildingService extends Service {

    static async getBuildings(filter : string, page : number, limit : number) {

        const response = await this.axios.get<Page<Building> | APIError>('/api/building', {
            params: {
                filter,
                page,
                limit
            }
        });

        return response.data;

    }

    static async getBuildingsByManager(managerId : number, filter : string, page : number, limit : number) {

        const response = await this.axios.get<Page<Building> | APIError>(`/api/building/manager/${managerId}`, {
            params: {
                filter,
                page,
                limit
            }
        });

        return response.data;

    }

    static async createBuilding(managerId : number, address : string) {

        const response = await this.axios.post<Building | APIError>('/api/building', {
            managerId,
            address
        });

        return response.data;

    }

    static async updateBuilding(id : number, managerId : optional<number>, address : optional<string>) {

        const response = await this.axios.put<Building | APIError>(`/api/building/${id}`, {
            managerId,
            address
        });

        return response.data;

    }

    static async deleteBuilding(id : number) {

        const response = await this.axios.delete<Building | APIError>(`/api/building/${id}`);

        return response.data;

    }

}